### PR TITLE
WIFSIGNALED means process has exited too (with a bang!)

### PR DESCRIPTION
**Pull request by arnetheduck**
for example test suite will hang if one of the tests crashes

### CODE REVIEW

**Comment by dom96** [comment]
Maybe that's why we get random Travis hangs... :)

**Comment by cheatfate** [comment]
Some of tests has crashes as expected behavior, and `testament` works properly.

This PR is not complete, because if `WIFSIGNALED` is true, then process exitcode becomes not reliable in current implementation, because OS stores signal number in place of exitcode. So this behavior must be described and/or resolved too.

**Comment by arnetheduck** [comment]
> Some of tests has crashes as expected behavior, and testament works properly.
not hard crashes like segfaults - see `execCmdEx` which loops on `peekExitCode` - without this patch, it will keep looping because peekExitCode will never notice the app is gone.

> This PR is not complete, because if WIFSIGNALED is true, then process exitcode becomes not reliable in current implementation, because OS stores signal number in place of exitcode. So this behavior must be described and/or resolved too.

good point. one way would be to go the way of the shell and return 128 + signal - this should be adequate given that this is what most unix apps do / expect (http://www.tldp.org/LDP/abs/html/exitcodes.html) and doesn't require any extensive api changes

**Comment by Araq** [comment]
This needs a test.

**Comment by arnetheduck** [comment]
uh, is there any precedent for such a test somewhere? ie which compiles an extra app and launches it? something to copy-paste from?

**Comment by cheatfate** [comment]
@arnetheduck you can start running program `yourself` one more time with arguments, which will cause to run other branch.

Also i'm curious, why you trying to split `posix.nim` for this PR? This PR is about `osproc.nim` issue, and not about `posix.nim` split. Also i dont think `posix.nim` must be splitted, because POSIX is standard for all platforms.

**Comment by arnetheduck** [comment]
@cheatfate good idea. 

this PR does not split posix more than it is already split (by platform) - posix does not specify how to get the return values other than saying that one should use the macros (`WIFEXITED`, `WEXITSTATUS` and friends).. so this impl simply uses platform specific knowledge to implement these macros the way it's done on linux, and uses the generic versions on other posix-like systems.

At least, that's my reading of http://pubs.opengroup.org/onlinepubs/9699919799/functions/wait.html

on that topic, there are many posix standards, which one does posix.nim implement?

**Comment by arnetheduck** [comment]
in fact, you can say that the current osproc is kind of wrong to assume `x&0xff00 shr 8`

File: lib/posix/posix_linux_amd64.nim [hunk 1]
```diff
@@ -602,3 +602,12 @@ var
 include posix_linux_amd64_consts
```
Comment (cheatfate) [inline lib/posix/posix_linux_amd64.nim]: There is no such file...

**Comment by cheatfate** [comment]
When i said **split**, i mean split one `posix.nim` file with different OS branches inside of it on 100 files like this:
```
posix_linux_i386.nim
posix_linux_m68k.nim
posix_linux_alpha.nim
posix_linux_powerpc.nim
posix_linux_powerpc64.nim
posix_linux_powerpc64el.nim
posix_linux_sparc.nim
posix_linux_vm.nim
posix_linux_ia64.nim
posix_linux_amd64.nim
posix_linux_mips.nim
posix_linux_mipsel.nim
posix_linux_arm.nim
posix_linux_arm64.nim
posix_linux_js.nim
posix_linux_nimrodvm.nim
posix_linux_avr.nim
posix_linux_msp430.nim
posix_linux_sparc64.nim
...
posix_freebsd_*.nim
...
posix_openbsd_*.nim
...
posix_dragonfly_*.nim
...
posix_netbsd_*.nim
...
```

But you representing only one `posix_linux_amd64.nim` and we both know, that you don't care about other OSes. So is this good solution? I dont think so.

`posix.nim` already has examples of branches for different posix related OSes.

**Comment by arnetheduck** [comment]
split is already part of `devel` - it was merged some time ago (?) - this branch merely adds on top of that to make it consistent - and more `posix` compatible for the `other` case where the c macros should be used

**Comment by cheatfate** [comment]
yeah i think i have missed that commit, but i still dont want to have such hell in `posix/` directory.
@Araq are you really want it?

**Comment by arnetheduck** [comment]
@cheatfate see https://git.musl-libc.org/cgit/musl/tree/arch for an example of split structure - they seem pretty happy about their decision to go the trivially simple way of split files, instead of complicating things with too many layers of `#ifdef` magic, and it's a joy to read / understand

**Comment by cheatfate** [comment]
For some reason i dont see any directories in `posix/` as you mentioned in `musl`. i see bunch of files...

**Comment by arnetheduck** [comment]
@cheatfate I don't mind directories with files in them, it boils down to pretty much the same thing.. or do you have any better ideas for dealing with large amounts of platform specific stuff? I'll be happy to change, it's not a big deal

**Comment by Araq** [comment]
I'm merging this, the refactoring of posix.nim into something that's nicer to work with is independent of this PR and should be it's own issue / discussionl

**Pull request merged**